tests/parser/base.py

```diff
@@ -88,21 +90,33 @@ def generate_from_keyfile(self, keyfile, netdef_id=None, expect_fail=False, file
                 if found_values >= 2:
                     break
             netdef_id = 'NM-' + uuid
-        if not filename:
-            filename = 'netplan-{}{}.nmconnection'.format(netdef_id, ssid)
-        f = os.path.join(self.workdir.name, 'run/NetworkManager/system-connections/{}'.format(filename))
+        generated_file = 'netplan-{}{}.nmconnection'.format(netdef_id, ssid)
+        original_file = filename or generated_file
+        f = os.path.join(self.workdir.name,
+                         'run/NetworkManager/system-connections/{}'.format(original_file))
         os.makedirs(os.path.dirname(f))
+        # Create the original keyfile that will be parsed by netplan
         with open(f, 'w') as file:
             file.write(keyfile)
 
         with capture_stderr() as outf:
             if expect_fail:
-                self.assertFalse(lib.netplan_parse_keyfile(f.encode(), None))
+                self.assertFalse(lib.netplan_parse_keyfile(f.encode(), ctypes.byref(err)))
+                if err:
+                    return err.contents.message.decode('utf-8')
             else:
-                self.assertTrue(lib.netplan_parse_keyfile(f.encode(), None))
+                self.assertTrue(lib.netplan_parse_keyfile(f.encode(), ctypes.byref(err)))
+                if err:  # pragma: nocover (this case should never happen)
```

Comment:
suggestion (non-blocking): the comment isn't exactly true, this codepath should actually happen, but isn't relevant for code coverage

```suggestion
                if err:  # pragma: nocover (only happens if a test fails so irrelevant for coverage)
```